File: c/bench.c
// CRDT Merge Micro-Benchmark — Pure C
// G-Counter merge: element-wise max over per-node counts
// OR-Set merge: union of constraint IDs with tombstone check
// Full state merge: composite of above

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <time.h>

#define N_NODES 32
#define N_CONSTRAINTS 256
#define N_RUNS 1000000

typedef struct {
    uint64_t counts[N_NODES];
    uint64_t violations[N_NODES];
} gcounter_t;

typedef struct {
    uint8_t active[N_CONSTRAINTS];      // 1 = active, 0 = removed
    uint8_t tombstones[N_CONSTRAINTS];  // 1 = removed at some point
} orset_t;

typedef struct {
    gcounter_t counter;
    orset_t constraints;
    int32_t pos_a;
    int32_t pos_b;
    int64_t norm;
    uint64_t version;
} state_t;

// ---- G-Counter merge: max per element ----
static inline void gcounter_merge(gcounter_t *dst, const gcounter_t *src) {
    for (int i = 0; i < N_NODES; i++) {
        if (src->counts[i] > dst->counts[i]) dst->counts[i] = src->counts[i];
        if (src->violations[i] > dst->violations[i]) dst->violations[i] = src->violations[i];
    }
}

// ---- OR-Set merge: union with tombstone removal ----
static inline void orset_merge(orset_t *dst, const orset_t *src) {
    for (int i = 0; i < N_CONSTRAINTS; i++) {
        // Merge tombstones
        if (src->tombstones[i]) dst->tombstones[i] = 1;
        // Merge active (union)
        if (src->active[i]) dst->active[i] = 1;
        // Remove active if tombstoned
        if (dst->tombstones[i]) dst->active[i] = 0;
    }
}

// ---- Full state merge ----
static inline void state_merge(state_t *dst, const state_t *src) {
    gcounter_merge(&dst->counter, &src->counter);
    orset_merge(&dst->constraints, &src->constraints);
    // Position: lower norm wins
    if (src->norm < dst->norm) {
        dst->pos_a = src->pos_a;
        dst->pos_b = src->pos_b;
        dst->norm = src->norm;
    }
    dst->version = (dst->version > src->version) ? dst->version : src->version;
}

// ---- Eisenstein norm ----
static inline int64_t e12_norm(int32_t a, int32_t b) {
    int64_t la = a, lb = b;
    return la * la - la * lb + lb * lb;
}

// ---- Bloom filter merge: bitwise OR on 64-bit words ----
#define BLOOM_WORDS 94  // ~6000 bits for 10K items at 1% FPR
typedef struct {
    uint64_t bits[BLOOM_WORDS];
} bloom_t;

static inline void bloom_merge(bloom_t *dst, const bloom_t *src) {
    for (int i = 0; i < BLOOM_WORDS; i++) {
        dst->bits[i] |= src->bits[i];
    }
}

// ---- Count-Min Sketch merge: element-wise max ----
#define SKETCH_D 7
#define SKETCH_W 1000
typedef struct {
    uint64_t counters[SKETCH_D][SKETCH_W];
} sketch_t;

static inline void sketch_merge(sketch_t *dst, const sketch_t *src) {
    for (int d = 0; d < SKETCH_D; d++) {
        for (int w = 0; w < SKETCH_W; w++) {
            if (src->counters[d][w] > dst->counters[d][w])
                dst->counters[d][w] = src->counters[d][w];
        }
    }
}

static uint64_t now_ns(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (uint64_t)ts.tv_sec * 1000000000ULL + (uint64_t)ts.tv_nsec;
}

#define BENCH(name, merge_fn, init_fn, type) do { \
    type a, b; \
    init_fn(&a); init_fn(&b); \
    uint64_t start = now_ns(); \
    for (int i = 0; i < N_RUNS; i++) { \
        merge_fn(&a, &b); \
    } \
    uint64_t elapsed = now_ns() - start; \
    double ns_per = (double)elapsed / N_RUNS; \
    printf("  %-30s %8.1f ns/op  %10.0f ops/s\n", name, ns_per, 1e9 / ns_per); \
} while(0)

static void gc_init(gcounter_t *g) { memset(g, 0, sizeof(*g)); g->counts[0] = 100; g->counts[5] = 200; }
static void ors_init(orset_t *o) { memset(o, 0, sizeof(*o)); for(int i=0;i<128;i++) o->active[i]=1; }
static void st_init(state_t *s) { memset(s, 0, sizeof(*s)); s->counter.counts[0]=100; s->norm=e12_norm(3,0); for(int i=0;i<64;i++) s->constraints.active[i]=1; }
static void bl_init(bloom_t *b) { memset(b, 0, sizeof(*b)); b->bits[0]=0xFF; b->bits[10]=0xAA; }
static void sk_init(sketch_t *s) { memset(s, 0, sizeof(*s)); s->counters[0][0]=100; s->counters[3][500]=200; }

int main(void) {
    printf("=== CRDT Merge Benchmark — C (gcc -O2) ===\n");
    printf("Runs: %d, Nodes: %d, Constraints: %d\n\n", N_RUNS, N_NODES, N_CONSTRAINTS);

    BENCH("G-Counter merge (32 nodes)", gcounter_merge, gc_init, gcounter_t);
    BENCH("OR-Set merge (256 constraints)", orset_merge, ors_init, orset_t);
    BENCH("Full state merge (64 active)", state_merge, st_init, state_t);
    BENCH("Bloom filter merge (6008 bits)", bloom_merge, bl_init, bloom_t);
    BENCH("Sketch merge (7×1000)", sketch_merge, sk_init, sketch_t);

    printf("\n--- Data structure sizes ---\n");
    printf("  gcounter_t: %zu bytes\n", sizeof(gcounter_t));
    printf("  orset_t:    %zu bytes\n", sizeof(orset_t));
    printf("  state_t:    %zu bytes\n", sizeof(state_t));
    printf("  bloom_t:    %zu bytes\n", sizeof(bloom_t));
    printf("  sketch_t:   %zu bytes\n", sizeof(sketch_t));

    return 0;
}

File: rust/bench.rs
// CRDT Merge Micro-Benchmark — Rust
// Same operations as C version for direct comparison

use std::time::Instant;

const N_NODES: usize = 32;
const N_CONSTRAINTS: usize = 256;
const BLOOM_WORDS: usize = 94;
const SKETCH_D: usize = 7;
const SKETCH_W: usize = 1000;
const N_RUNS: usize = 1_000_000;

struct GCounter {
    counts: [u64; N_NODES],
    violations: [u64; N_NODES],
}

struct ORSet {
    active: [u8; N_CONSTRAINTS],
    tombstones: [u8; N_CONSTRAINTS],
}

struct BloomFilter {
    bits: [u64; BLOOM_WORDS],
}

struct Sketch {
    counters: [[u64; SKETCH_W]; SKETCH_D],
}

fn gcounter_merge(dst: &mut GCounter, src: &GCounter) {
    for i in 0..N_NODES {
        dst.counts[i] = dst.counts[i].max(src.counts[i]);
        dst.violations[i] = dst.violations[i].max(src.violations[i]);
    }
}

fn orset_merge(dst: &mut ORSet, src: &ORSet) {
    for i in 0..N_CONSTRAINTS {
        if src.tombstones[i] != 0 { dst.tombstones[i] = 1; }
        if src.active[i] != 0 { dst.active[i] = 1; }
        if dst.tombstones[i] != 0 { dst.active[i] = 0; }
    }
}

fn bloom_merge(dst: &mut BloomFilter, src: &BloomFilter) {
    for i in 0..BLOOM_WORDS {
        dst.bits[i] |= src.bits[i];
    }
}

fn sketch_merge(dst: &mut Sketch, src: &Sketch) {
    for d in 0..SKETCH_D {
        for w in 0..SKETCH_W {
            dst.counters[d][w] = dst.counters[d][w].max(src.counters[d][w]);
        }
    }
}

fn bench<F: FnMut()>(name: &str, mut f: F) {
    // Warmup
    for _ in 0..1000 { f(); }
    
    let start = Instant::now();
    for _ in 0..N_RUNS { f(); }
    let elapsed = start.elapsed();
    let ns_per = elapsed.as_nanos() as f64 / N_RUNS as f64;
    println!("  {:<35} {:>8.1} ns/op  {:>10.0} ops/s", name, ns_per, 1e9 / ns_per);
}

fn main() {
    println!("=== CRDT Merge Benchmark — Rust (rustc -O2, Zen 5) ===");
    println!("Runs: {}, Nodes: {}, Constraints: {}\n", N_RUNS, N_NODES, N_CONSTRAINTS);

    let mut gc_a = GCounter { counts: [0; N_NODES], violations: [0; N_NODES] };
    let mut gc_b = GCounter { counts: [0; N_NODES], violations: [0; N_NODES] };
    gc_a.counts[0] = 100; gc_b.counts[5] = 200;

    let mut ors_a = ORSet { active: [0; N_CONSTRAINTS], tombstones: [0; N_CONSTRAINTS] };
    let mut ors_b = ORSet { active: [0; N_CONSTRAINTS], tombstones: [0; N_CONSTRAINTS] };
    for i in 0..128 { ors_a.active[i] = 1; }

    let mut bl_a = BloomFilter { bits: [0; BLOOM_WORDS] };
    let mut bl_b = BloomFilter { bits: [0; BLOOM_WORDS] };
    bl_a.bits[0] = 0xFF; bl_b.bits[10] = 0xAA;

    let mut sk_a = Sketch { counters: [[0u64; SKETCH_W]; SKETCH_D] };
    let mut sk_b = Sketch { counters: [[0u64; SKETCH_W]; SKETCH_D] };
    sk_a.counters[0][0] = 100; sk_b.counters[3][500] = 200;

    bench("G-Counter merge (32 nodes)", || { gcounter_merge(&mut gc_a, &gc_b); });
    bench("OR-Set merge (256 constraints)", || { orset_merge(&mut ors_a, &ors_b); });
    bench("Bloom filter merge (752 bytes)", || { bloom_merge(&mut bl_a, &bl_b); });
    bench("Sketch merge (7×1000)", || { sketch_merge(&mut sk_a, &sk_b); });

    // SIMD-style unrolled Bloom merge
    bench("Bloom merge (unrolled ×4)", || {
        let src = &bl_b.bits;
        let dst = &mut bl_a.bits;
        let mut i = 0;
        while i + 3 < BLOOM_WORDS {
            dst[i] |= src[i]; dst[i+1] |= src[i+1];
            dst[i+2] |= src[i+2]; dst[i+3] |= src[i+3];
            i += 4;
        }
        while i < BLOOM_WORDS { dst[i] |= src[i]; i += 1; }
    });

    // Iterator-based Bloom merge
    bench("Bloom merge (iterator zip)", || {
        bl_a.bits.iter_mut().zip(bl_b.bits.iter()).for_each(|(d, s)| *d |= *s);
    });

    // Unsafe raw pointer Bloom merge
    bench("Bloom merge (unsafe ptr)", || {
        unsafe {
            let dst = bl_a.bits.as_mut_ptr();
            let src = bl_b.bits.as_ptr();
            for i in 0..BLOOM_WORDS {
                *dst.add(i) |= *src.add(i);
            }
        }
    });

    println!("\n--- Data structure sizes ---");
    println!("  GCounter:    {} bytes", std::mem::size_of::<GCounter>());
    println!("  ORSet:       {} bytes", std::mem::size_of::<ORSet>());
    println!("  BloomFilter: {} bytes", std::mem::size_of::<BloomFilter>());
    println!("  Sketch:      {} bytes", std::mem::size_of::<Sketch>());
}
